### PR TITLE
Feature/cfd 358 allow bypass for unmatched farestages

### PR DESCRIPTION
**Pull request by Aaron-Collis-IW**
# Description

Allow users to use a checkbox to bypass the requirement to map all fare stages to at least one stop. This does not allow the user to map zero fare stages.

# Testing instructions

- Go through a single ticket journey to the point where you have to map farestages.
- Try to continue with 0 fare stages mapped.
- Try to continue with 1 stage mapped.
- Try to continue with 1 stage mapped and bypass checkbox ticked.

- Try this again with any of the following: Single, Return, P2P Period, Single/Return Carnets

# Type of change

-   [x] feat - A new feature
-   [ ] fix - A bug fix
-   [ ] docs - Documentation only changes
-   [ ] codestyle - Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
-   [ ] refactor - A code change that neither fixes a bug nor adds a feature
-   [ ] perf - A code change that improves performance
-   [ ] test - Adding missing tests or correcting existing tests
-   [ ] build - Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm)
-   [ ] ci - Changes to our CI configuration files and scripts (example scopes: Travis, Circle, BrowserStack, SauceLabs)
-   [ ] chore - Other changes that don't modify src or test files
-   [ ] revert - Reverts a previous commit
-   [ ] content - Changes to content or copy

### CODE REVIEW

File: repos/fdbt-site/cypress_tests/cypress/integration/period.ts [hunk 1]
```diff
@@ -49,7 +49,15 @@ describe('the period faretype product journey', () => {
     it('completes successfully for point to point period products', () => {
         selectFareType('period', false);
         defineUserTypeAndTimeRestrictions();
-        completePointToPointPeriodPages();
+        completePointToPointPeriodPages(false);
+        completeSalesPages();
+        isUuidStringValid();
+    });
+
+    it.only('completes successfully for point to point period products with partial fare stage matching', () => {
```
Comment (jacsamell) [inline repos/fdbt-site/cypress_tests/cypress/integration/period.ts:57]: Remove the .only

File: repos/fdbt-site/src/components/WarningSummary.tsx [hunk 2]
```diff
@@ -0,0 +1,47 @@
+import React, { ReactElement } from 'react';
+import { ErrorInfo } from '../interfaces';
+
+interface ErrorSummary {
```
Comment (jacsamell) [inline repos/fdbt-site/src/components/WarningSummary.tsx:4]: Assume the places in here that say Error should say Warning?

File: repos/fdbt-site/src/components/WarningSummary.tsx [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+                        <div className="govuk-checkboxes__item">
+                            <input
+                                className="govuk-checkboxes__input"
+                                id="bypass"
+                                name="overrideWarning"
+                                value="yes"
+                                type="checkbox"
+                            />
+                            <label htmlFor="overrideWarning" className="govuk-label govuk-checkboxes__label">
+                                Check this box if you wish to proceed without assigning all fare stages, then click
```
Comment (jacsamell) [inline repos/fdbt-site/src/components/WarningSummary.tsx:36]: Should this message be passed in as a props?

File: repos/fdbt-site/src/components/WarningSummary.tsx [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+                {errors[0].errorMessage}
+            </h2>
+            <div className="govuk-error-summary__body">
+                <ul className="govuk-list govuk-error-summary__list">
+                    <div className="govuk-checkboxes">
+                        <div className="govuk-checkboxes__item">
+                            <input
+                                className="govuk-checkboxes__input"
+                                id="bypass"
+                                name="overrideWarning"
```
Comment (jacsamell) [inline repos/fdbt-site/src/components/WarningSummary.tsx:31]: Should this be passed in as a props?

File: repos/fdbt-site/tests/components/MatchingBase.test.tsx [hunk 5]
```diff
@@ -121,6 +123,7 @@ describe('MatchingBase', () => {
                         stops={zoneStops}
                         service={service}
                         error={false}
+                        warning={false}
```
Comment (jacsamell) [inline repos/fdbt-site/tests/components/MatchingBase.test.tsx:126]: Do we need some tests in here where warning is set to true?

File: repos/fdbt-site/cypress_tests/cypress/support/steps.ts [hunk 6]
```diff
@@ -169,12 +184,17 @@ const completePointToPointPeriodProductDetail = (): void => {
...
 
-export const completeSinglePages = (csvUpload: boolean, isCarnet: boolean): void => {
+export const completeSinglePages = (csvUpload: boolean, isCarnet: boolean, partialMatching: boolean): void => {
     completeServicePage();
     selectRandomOptionFromDropDown('direction-journey-pattern');
     continueButtonClick();
     completeFareTrianglePages(csvUpload);
-    completeMatchingPage();
+
+    if (partialMatching) {
```
Comment (jacsamell) [inline repos/fdbt-site/cypress_tests/cypress/support/steps.ts:193]: As discussed if this doesn't work and we are making it in this ticket then should we remove the changes to the UI tests?

File: repos/fdbt-site/tests/pages/api/inboundMatching.test.ts [hunk 7]
```diff
@@ -61,7 +61,7 @@ describe('Inbound Matching API', () => {
 
     it('correctly generates matching error info, updates the INBOUND_MATCHING_ATTRIBUTE and then redirects to inboundMatching page when there are unassigned fare stages', () => {
         const mockMatchingError: MatchingWithErrors = {
-            error: true,
+            warning: true,
```
Comment (jacsamell) [inline repos/fdbt-site/tests/pages/api/inboundMatching.test.ts:64]: Should there be a test for error here as well? The old test was still valid I think?

File: repos/fdbt-site/tests/pages/api/matching.test.ts [hunk 8]
```diff
@@ -87,7 +87,7 @@ describe('Matching API', () => {
 
     it('correctly generates matching error info, updates the MATCHING_ATTRIBUTE and then redirects to matching page when there are unassigned fare stages', () => {
         const mockMatchingError: MatchingWithErrors = {
-            error: true,
+            warning: true,
```
Comment (jacsamell) [inline repos/fdbt-site/tests/pages/api/matching.test.ts:90]: Same here as above, think the test for error is still valid?

File: repos/fdbt-site/tests/pages/api/outboundMatching.test.ts [hunk 9]
```diff
@@ -62,7 +62,7 @@ describe('Outbound Matching API', () => {
 
     it('correctly generates matching error info, updates the MATCHING_ATTRIBUTE and then redirects to outboundMatching page when there are unassigned fare stages', () => {
         const mockMatchingError: MatchingWithErrors = {
-            error: true,
+            warning: true,
```
Comment (jacsamell) [inline repos/fdbt-site/tests/pages/api/outboundMatching.test.ts:65]: And here

**Review by jacsamell — CHANGES_REQUESTED** [review]

File: repos/fdbt-site/src/components/MatchingBase.tsx [hunk 10]
```diff
@@ -174,14 +178,17 @@ const MatchingBase = ({
     };
 
     if (error) {
-        errors.push({ errorMessage: 'Ensure each fare stage is assigned at least once.', id: 'option-0' });
+        errors.push({ errorMessage: 'No fare stages have been assigned.', id: 'option-0' });
```
Comment (RobCatton) [inline repos/fdbt-site/src/components/MatchingBase.tsx:181]: I think we should have an error message which advises the user on how they can fix their error, whilst also telling them what they have done wrong.

Comment (Aaron-Collis-IW) [reply repos/fdbt-site/src/components/MatchingBase.tsx:181]: So just "No fare stages have been assigned, please assign each fare stage to a stop."

File: repos/fdbt-site/tests/pages/api/inboundMatching.test.ts (hunk 7 above)
Comment (Aaron-Collis-IW) [reply repos/fdbt-site/tests/pages/api/inboundMatching.test.ts:64]: having trouble getting this to work :/ 

I'm able to set the expected data to error but unsure what I need to feed in the get the actual to output an error

File: repos/fdbt-site/src/components/WarningSummary.tsx (hunk 3 above)
Comment (Aaron-Collis-IW) [reply repos/fdbt-site/src/components/WarningSummary.tsx:36]: As the warning message is only using for the matching pages would it be btter to just leave it as is and rename the file to MatchingWarning?

File: repos/fdbt-site/src/components/MatchingBase.tsx (hunk 10 above)
Comment (jacsamell) [reply repos/fdbt-site/src/components/MatchingBase.tsx:181]: Did you fix this @Aaron-Collis-IW ?

File: repos/fdbt-site/tests/pages/api/inboundMatching.test.ts [hunk 11]
```diff
@@ -59,11 +59,32 @@ describe('Inbound Matching API', () => {
         expect(writeHeadMock).toBeCalledWith(302, { Location: '/returnValidity' });
     });
 
-    it('correctly generates matching error info, updates the INBOUND_MATCHING_ATTRIBUTE and then redirects to inboundMatching page when there are unassigned fare stages', () => {
+    it('correctly generates matching warning info, updates the INBOUND_MATCHING_ATTRIBUTE and then redirects to inboundMatching page when there are unassigned fare stages', () => {
```
Comment (jacsamell) [inline repos/fdbt-site/tests/pages/api/inboundMatching.test.ts:62]: Minor think you have the warning and error the wrong way round in these 2 tests

**Review by jacsamell — APPROVED** [review]

**Pull request merged**